runtime/test/strategies/coalesce-recipes-test.js

```diff
@@ -0,0 +1,109 @@
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+'use strict';
+
+import {Manifest} from '../../manifest.js';
+import {StrategyTestHelper} from './strategy-test-helper.js';
+import {CoalesceRecipes} from '../../strategies/coalesce-recipes.js';
+import {assert} from '../chai-web.js';
+
+describe('CoalesceRecipes', function() {
+  it('coalesces required slots', async () => {
+    let manifest = (await Manifest.parse(`
+      particle P1
+        consume root
+          must provide foo
+
+      particle P2
+        consume foo
+
+      recipe
+        slot 'id0' as slot0
+        P1
+          consume root as slot0
+      recipe
+        P2
+    `));
+
+    let recipes = manifest.recipes;
+    recipes.forEach(recipe => {
+      recipe.normalize();
+      assert.isFalse(recipe.isResolved());
+    });
+    let arc = StrategyTestHelper.createTestArc('test-plan-arc', manifest, 'dom');
+    let inputParams = {generated: [], terminal: [{result: recipes[0], score: 1}, {result: recipes[1], score: 1}]};
+    let strategy = new CoalesceRecipes(arc);
+    let results = await strategy.generate(inputParams);
+
+    assert.equal(1, results.length);
+    let recipe = results[0].result;
+    assert.isTrue(recipe.isResolved());
+    assert.lengthOf(recipe.particles, 2);
+    assert.lengthOf(recipe.slots, 2);
+  });
+
+  it('coalesces required slots with handles', async () => {
+    let manifest = (await Manifest.parse(`
+      schema Thing
+      schema OtherThing
+      particle P1
+        in Thing thing
+        consume root
+          must provide foo
+            handle thing
+
+      particle P2
+        in Thing thing
+        out OtherThing other
+        consume foo
+
+      particle P3
+        out Thing thing
+
+      recipe
+        slot 'id0' as slot0
+        copy 'mything' as thingHandle
+        P1
+          thing = thingHandle
+          consume root as slot0
+
+      recipe
+        use as thingHandle
+        create as otherHandle
+        P2
+          thing = thingHandle
+          other = otherHandle
+        P3
+          thing = thingHandle
+      
+      resource MyThing
+          start
+          []
+      store Store0 of Thing 'mything' in MyThing
+    `));
+
+    let recipes = manifest.recipes;
+    recipes.forEach(recipe => {
+      recipe.normalize();
+      assert.isFalse(recipe.isResolved());
+    });
+    let arc = StrategyTestHelper.createTestArc('test-plan-arc', manifest, 'dom');
+    let inputParams = {generated: [], terminal: [{result: recipes[0], score: 1}, {result: recipes[1], score: 1}]};
+    let strategy = new CoalesceRecipes(arc);
+    let results = await strategy.generate(inputParams);
+
+    assert.equal(1, results.length);
+    let recipe = results[0].result;
+    recipe.normalize();
+    assert.isTrue(recipe.isResolved());
+    assert.lengthOf(recipe.particles, 3);
+    assert.lengthOf(recipe.slots, 2);
+  });
```

Comment:
I'm thinking what handle fates on the other side we should accept. E.g. in the below example:

```
      schema Thing

      particle P1
        in Thing thing
        consume root
          must provide foo
            handle thing

      particle P2
        in Thing thing
        consume foo

      recipe A
        slot 'id0' as slot0
        use as thingHandle
        P1
          thing = thingHandle
          consume root as slot0

      recipe B
        map as thingHandle
        P2
          thing = thingHandle
```

I don't think we should allow coalescing, as recipe B is explicit that it wants thingHandle mapped - i.e. semantically "find me THE BEST single candidate for this type/name/tags you can find" (AFAIU the vision for `map`), so coalescing and giving B whatever A uses is not ok. I would argue that we should only allow 'use' or '?' in B recipe.

WDYT?

Reply:
I think the following cases should be allowed:
- fates 'use' or '?' for B work with any fate in A's handle (if the directions allow)
- 'map' or 'copy' in B work with 'map' or 'copy' in A (if the directions allow)
- and 'create' in B works with 'create' in 'A'

interesting, what is A's handle fates is '?'?

WDYT?

Reply:
Good points about matching map/copy between A and B. I would say (for a start, we can revisit) that if A is '?' and B is 'map' than don't allow, because we don't know what the fate will end up being. If A is '?' than B has to be 'use' or '?'. What do you think?